third_party/move/move-vm/runtime/src/loader/mod.rs

```diff
@@ -114,61 +123,281 @@ lazy_static! {
+        matches!(self, Self::V1(loader) if *loader.invalidated.read())
+    }
+
+    pub(crate) fn check_script_dependencies_and_check_gas(
+        &self,
+        module_store: &ModuleStorageAdapter,
+        data_store: &mut TransactionDataCache,
+        gas_meter: &mut impl GasMeter,
+        traversal_context: &mut TraversalContext,
+        script_blob: &[u8],
+        module_storage: &impl ModuleStorageV2,
+        script_storage: &impl ScriptStorage,
+    ) -> VMResult<()> {
+        match self {
+            Self::V1(loader) => loader.check_script_dependencies_and_check_gas(
+                module_store,
+                data_store,
+                gas_meter,
+                traversal_context,
+                script_blob,
+            ),
+            Self::V2(loader) => loader
+                .check_script_dependencies_and_check_gas(
+                    module_storage,
+                    script_storage,
+                    gas_meter,
+                    traversal_context,
+                    script_blob,
+                )
+                .map_err(|e| e.finish(Location::Undefined)),
         }
     }
 
-    pub(crate) fn insert_or_get(&self, name: StructIdentifier) -> StructNameIndex {
-        if let Some(idx) = self.data.read().0.get(&name) {
-            return *idx;
+    pub(crate) fn check_dependencies_and_charge_gas<'a, I>(
+        &self,
+        module_store: &ModuleStorageAdapter,
+        data_store: &mut TransactionDataCache,
+        gas_meter: &mut impl GasMeter,
+        visited: &mut BTreeMap<(&'a AccountAddress, &'a IdentStr), ()>,
+        referenced_modules: &'a Arena<Arc<CompiledModule>>,
+        ids: I,
+        module_storage: &dyn ModuleStorageV2,
+    ) -> VMResult<()>
+    where
+        I: IntoIterator<Item = (&'a AccountAddress, &'a IdentStr)>,
+        I::IntoIter: DoubleEndedIterator,
+    {
+        match self {
+            Self::V1(loader) => loader.check_dependencies_and_charge_gas(
+                module_store,
+                data_store,
+                gas_meter,
+                visited,
+                referenced_modules,
+                ids,
+            ),
+            Self::V2(loader) => loader
+                .check_dependencies_and_charge_gas(
+                    module_storage,
+                    gas_meter,
+                    visited,
+                    referenced_modules,
+                    ids,
+                )
+                .map_err(|e| e.finish(Location::Undefined)),
         }
-        let mut inner_data = self.data.write();
-        let idx = StructNameIndex(inner_data.1.len());
-        inner_data.0.insert(name.clone(), idx);
-        inner_data.1.push(name);
-        idx
     }
 
-    pub(crate) fn idx_to_identifier(
+    pub(crate) fn load_script(
         &self,
-        idx: StructNameIndex,
+        script_blob: &[u8],
+        ty_args: &[TypeTag],
+        data_store: &mut TransactionDataCache,
+        module_store: &ModuleStorageAdapter,
```

Comment:
This api looks very messy. I think I see your point of creating a versioned loader but was wondering what's that gonna buy us asides from being able to choose between which loader to use?

Reply:
The idea is to remove V1 completely. It should be possible I hope module a few cases where module metadata is messed up.

Reply:
Also note that transaction data cache argument will be removed as scripts will be cached outside in V2 design. The only reason it is here is because we cache scripts there when we traverse dependencies 🤦‍♂️

Reply:
Working on a separate branch, so let's keep it as is for now